### DISCUSSION TITLE
CPU usage is max even when no ingestion is taking place and with small amount of data

### DISCUSSION THREAD

**Discussion by sarim2000** [pathwaycom/llm-app#78] [General]
Any solution for this, am reading from local directory, md format data, binary and streaming mode

**Comment by berkecanrizai** [comment]
Hey @sarim2000 , could you tell about your hardware, which app you are running (one of the provided ones or a custom one), whether you are running on Docker, Pathway version, and the modules you are using (such as parser, splitter, OpenAI embedder, etc.)? 
Thanks,

**Reply by sarim2000** [reply]
ec2 machine with with 4 cores and 16 gb ram 


On ec2 I first ran just python script and it worked till it keeping hitting cpu usage thus moved to docker, same issue docker restarting (used dockerfile from ur deployment example)


my code
```python
# Import necessary libraries and modules
from urllib.parse import urlparse
from pathway.udfs import DiskCache
import logging
import os
import pathway as pw
from dotenv import load_dotenv
from pathway.xpacks.llm.embedders import OpenAIEmbedder
from pathway.xpacks.llm.parsers import ParseUnstructured, ParseUtf8
from pathway.xpacks.llm.splitters import TokenCountSplitter
from pathway.xpacks.llm.vector_store import VectorStoreServer
import re
from datetime import datetime

# Load environment variables
load_dotenv()


def write_error_metadata_files(extracted_date, doc_name, timestamp):
    """
    Write the error metadata files to a csv

    Args:
        extracted_date (str): The date to write to the csv
        doc_name (str): The name of the document to write to the csv
        timestamp (int): The timestamp to write to the csv
    """
    with open(f"logs/error_files.csv", "a+") as f:
        f.write(f"{extracted_date},{doc_name},{timestamp}\n")


def write_success_metadata_files(extracted_date: str, doc_name: str, timestamp: int):
    """
    Write the success metadata files to a csv

    Args:
        extracted_date (str): The date to write to the csv
        doc_name (str): The name of the document to write to the csv
        timestamp (int): The timestamp to write to the csv
    """
    with open(f"logs/success_files.csv", "a+") as f:
        f.write(f"{extracted_date},{doc_name},{timestamp}\n")


def extract_domain(url):
    parsed_url = urlparse(url)
    domain = parsed_url.netloc.split(".")
    # Extract the domain name (usually the second last part of the netloc)
    return domain[-2] if len(domain) > 1 else domain[0]


def extract_published_date(file_text: str, metadata: dict) -> tuple[str, dict]:
    """
    Extract the published date from a document

    Args:
        file_text (str): The text of the document to extract the published date from
        metadata (dict): The metadata of the document to extract the published date from

    Returns:
        tuple: The text of the document and the metadata of the document
    """
    match = re.search(r"published_date:\s*(\d+)", file_text)
    source_url = re.search(r'source_url:\s*{"""(\S+)"""}', file_text)
    if match:
        unix_timestamp = int(match.group(1))
        try:
            # Validate the timestamp
            if unix_timestamp <= 0:
                raise ValueError("Timestamp is not valid")
            parsed_date = datetime.fromtimestamp(unix_timestamp)
            print(parsed_date)
            metadata["published_date"] = unix_timestamp

            write_success_metadata_files(
                parsed_date.isoformat(), metadata["path"], metadata["published_date"]
            )
        except (ValueError, TypeError) as e:
            write_error_metadata_files(
                str(unix_timestamp), metadata["path"], ValueError
            )
            metadata["published_date"] = 0
    else:
        write_error_metadata_files("No match", metadata["path"], 0)
        metadata["published_date"] = 0

    if source_url:
        metadata["publisher"] = extract_domain(source_url.group(1))
        print("publisher", metadata["publisher"])
        write_success_metadata_files(
            metadata["publisher"], metadata["path"], source_url.group(1)
        )

    else:
        write_error_metadata_files("No source url", metadata["path"], 0)
        metadata["publisher"] = "unknown"
    return file_text, metadata


# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(name)s %(levelname)s %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)

# Initialize data sources list
data_sources = []
# Append file system reader to data sources
data_sources.append(
    pw.io.fs.read(
        path="./content",
        format="binary",
        mode="streaming",
        with_metadata=True,
        persistent_id="content",
    )
)

# Initialize text splitter and embedder
text_splitter = TokenCountSplitter()
embedder = OpenAIEmbedder(
    api_key=os.getenv("OPENAI_API_KEY"),
    cache_strategy=DiskCache(),
)

# Setup vector store server with the initialized components
vector_server = VectorStoreServer(
    *data_sources,
    embedder=embedder,
    splitter=text_splitter,
    parser=ParseUtf8(),
    doc_post_processors=[extract_published_date],
)


# Run the vector store server
vector_server.run_server(host="127.0.0.1", port=8767, threaded=False, with_cache=True)

# persistence_backend = pw.persistence.Backend.filesystem("./Cache/")
# persistence_config = pw.persistence.Config.simple_config(persistence_backend)
```

my data sources are .md (currently its 17k and counting)

**Reply by sarim2000** [reply]
@berkecanrizai  any solution for this? 

I even thought doing ingestion in other device moving cache folder to main ec2 machine but even then cpu limits are being hit, and we dont have any other process that is hogging that machine, pathway is hitting it to max even when not ingesting

**Reply by berkecanrizai** [reply]
@sarim2000 I think this is very likely the result of io writes in the `UDF` (user defined function). The functions you put into `doc_post_processors` are converted to `pw.UDF`s and applied to tables inside of the VectorStoreServer.

File writes are supported with different library modules and output connectors, [CSV writer](https://pathway.com/developers/user-guide/connect/connectors/csv_connectors/) and [pw.io.subscribe](https://pathway.com/developers/api-docs/pathway-io#pathway.io.subscribe) are probably what you are looking for. 

You can use these on the parsed docs table that is inside of the vector store. To access that, you can use: `vector_server._graph["parsed_docs"]`.

I briefly checked your app, I think you can return a tuple from the function (string, filename), then store them in two separate columns. Later ingest/write to the file with `pw.io.subscribe` based on row values.

**Reply by sarim2000** [reply]
> I briefly checked your app, I think you can return a tuple from the function (string, filename), then store them in two separate columns. Later ingest/write to the file with pw.io.subscribe based on row values.

I did not get this point. Can u show me an example?

**Reply by sarim2000** [reply]
>  I think this is very likely the result of io writes in the UDF (user defined function). The functions you put into doc_post_processors are converted to pw.UDFs and applied to tables inside of the VectorStoreServer.

and why is this given as an option if it creates this issue? am i using it wrong?

**Reply by sarim2000** [reply]
more specifically, taking this rag example, can you tell me the flow so this doesnt cause max cpu usage..?

I did not get your answer. More specifically the flow, what should be the flow? As I am using pw.io.fs.read, should I make the changes here to add additional metadata column using pw.io.csv and then add vector server and not pass doc_post_processors in it?

**Reply by sarim2000** [reply]
also i remove writing to file inside doc_post_processor functions, its still hogging cpu resource, I am just extracting metadata adn saving

**Reply by berkecanrizai** [reply]
If this is happening even after removing the file save, this is not expected, I will try to replicate and get back to you.
That much CPU usage is not normal without any io write in the `UDF`s.

> I did not get this point. Can u show me an example?

I meant something like this (while removing the `doc_post_processors` and keeping the rest of the code same):

```python
def save_file_callback(key, row, time, is_addition):
    file_name = row["filename"].value
    write_success_metadata_files(row["date"], filename)

@pw.udf
def extract_published_date(doc) -> tuple[str, str]:
  data: dict = doc.as_dict()
  text = data["text"]
  metadata = data["metadata"]
  file_name = metadata["path"]
  ...
  return parsed_date, file_name

docs_table = vector_server._graph["parsed_docs"]

docs_table += docs_table.select(extract_tup=extract_published_date(pw.this.data))

result_table = docs_table.select(date=pw.this.extract_tup[0], filename=pw.this.extract_tup[1])

pw.io.subscribe(result_table, on_change=save_file_callback)
```

**Comment by KamilPiechowiak** [comment]
Hey @sarim2000,
thank you for your report. To know if it is a problem on the pathway side please make an experiment not involving pathway. You can read the files one by one and apply your processing logic and see how much time it takes then.
You can just read all files in your `./content` directory and apply `extract_published_date` to each of them.

**Reply by sarim2000** [reply]
even after removing doc_post_processor functions, after all ingestion pathway still takes up 130% on average

**Reply by KamilPiechowiak** [reply]
Ok. Thanks for info. I'll investigate it and will get back to you once I have some conclusions.

**Reply by sarim2000** [reply]
ya cool; buzz me if you guys need additional info

**Reply by KamilPiechowiak** [reply]
Hey @sarim2000, one fix is already there. If you install `pathway v0.13.2`, the cpu usage in idle state should be decreased significantly (in our tests it's on 4% now). We're working on decreasing it even further but hope that this fix will already make the app usable for you.
